src/libsystemd/sd-journal/journal-file.c

```diff
@@ -179,95 +178,23 @@ int journal_file_tail_end_by_mmap(JournalFile *f, uint64_t *ret_offset) {
-        bool wait = true;
-
+int journal_file_set_state(JournalFile *f, uint8_t state) {
         assert(f);
+        assert(state < _STATE_MAX);
 
         if (!journal_file_writable(f))
                 return -EPERM;
 
-        if (f->fd < 0 || !f->header)
+        if (!f->header)
                 return -EINVAL;
 
-        while (wait) {
-                switch (f->offline_state) {
-                case OFFLINE_JOINED:
-                        /* No offline thread, no need to wait. */
-                        wait = false;
-                        break;
-
-                case OFFLINE_SYNCING: {
-                                OfflineState tmp_state = OFFLINE_SYNCING;
-                                if (!__atomic_compare_exchange_n(&f->offline_state, &tmp_state, OFFLINE_CANCEL,
-                                    false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                                        continue;
-                        }
-                        /* Canceled syncing prior to offlining, no need to wait. */
-                        wait = false;
-                        break;
-
-                case OFFLINE_AGAIN_FROM_SYNCING: {
-                                OfflineState tmp_state = OFFLINE_AGAIN_FROM_SYNCING;
-                                if (!__atomic_compare_exchange_n(&f->offline_state, &tmp_state, OFFLINE_CANCEL,
-                                    false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                                        continue;
-                        }
-                        /* Canceled restart from syncing, no need to wait. */
-                        wait = false;
-                        break;
-
-                case OFFLINE_AGAIN_FROM_OFFLINING: {
-                                OfflineState tmp_state = OFFLINE_AGAIN_FROM_OFFLINING;
-                                if (!__atomic_compare_exchange_n(&f->offline_state, &tmp_state, OFFLINE_CANCEL,
-                                    false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                                        continue;
-                        }
-                        /* Canceled restart from offlining, must wait for offlining to complete however. */
-                        _fallthrough_;
-                default: {
-                        int r;
-
-                        r = journal_file_set_offline_thread_join(f);
-                        if (r < 0)
-                                return r;
-
-                        wait = false;
-                        break;
-                }
-                }
-        }
-
         if (mmap_cache_fd_got_sigbus(f->cache_fd))
                 return -EIO;
 
         switch (f->header->state) {
                 case STATE_ONLINE:
-                        return 0;
-
                 case STATE_OFFLINE:
-                        f->header->state = STATE_ONLINE;
-                        (void) fsync(f->fd);
+                        f->header->state = state;
```

Comment:
So when an append occurs you're just assigning the state regardless of what's going on right now?  Is there not an asynchronous `f->header->state=STATE_OFFLINE` when idle potentially occurring anymore?

Maybe I'm missing something, but my knee-jerk reaction right now is that this PR is entirely misguided.